sycl/source/detail/kernel_data.cpp

```diff
@@ -0,0 +1,351 @@
+    // So we just suppose that WG size is always default for stream.
+    // TODO adjust MNDRDesc when device image contains kernel's attribute
+    if (GlobalSize == 0) {
+      GlobalSize = NDRDesc.NumWorkGroups[0];
+      for (size_t I = 1; I < NDRDesc.Dims; ++I) {
+        GlobalSize *= NDRDesc.NumWorkGroups[I];
+      }
+    }
+    addArgsForGlobalAccessor(GFlushReq, Index, IndexShift, Size,
+                             IsKernelCreatedFromSource, GlobalSize, MArgs,
+                             MKernelIsESIMD);
+    ++IndexShift;
+    addArg(kernel_param_kind_t::kind_std_layout, &S->FlushBufferSize,
+           sizeof(S->FlushBufferSize), Index + IndexShift);
+
+    break;
+  }
+  case kernel_param_kind_t::kind_accessor: {
+    // For args kind of accessor Size is information about accessor.
+    // The first 11 bits of Size encodes the accessor target.
+    const access::target AccTarget =
+        static_cast<access::target>(Size & AccessTargetMask);
+    switch (AccTarget) {
+    case access::target::device:
+    case access::target::constant_buffer: {
+      detail::Requirement *AccImpl = static_cast<detail::Requirement *>(Ptr);
+      addArgsForGlobalAccessor(AccImpl, Index, IndexShift, Size,
+                               IsKernelCreatedFromSource, GlobalSize, MArgs,
+                               MKernelIsESIMD);
+      break;
+    }
+    case access::target::local: {
+      detail::LocalAccessorImplHost *LAccImpl =
+          static_cast<detail::LocalAccessorImplHost *>(Ptr);
+
+      addArgsForLocalAccessor(LAccImpl, Index, IndexShift,
+                              IsKernelCreatedFromSource, MArgs, MKernelIsESIMD);
+      break;
+    }
+    case access::target::image:
+    case access::target::image_array: {
+      detail::Requirement *AccImpl = static_cast<detail::Requirement *>(Ptr);
+      addArg(Kind, AccImpl, Size, Index + IndexShift);
+      if (!IsKernelCreatedFromSource) {
+        // TODO Handle additional kernel arguments for image class
+        // if the compiler front-end adds them.
+      }
+      break;
+    }
+    case access::target::host_image:
+    case access::target::host_task:
+    case access::target::host_buffer: {
+      throw sycl::exception(make_error_code(errc::invalid),
+                            "Unsupported accessor target case.");
+      break;
+    }
+    }
+    break;
+  }
+  case kernel_param_kind_t::kind_dynamic_accessor: {
+    const access::target AccTarget =
+        static_cast<access::target>(Size & AccessTargetMask);
+    switch (AccTarget) {
+    case access::target::local: {
+
+      // We need to recover the inheritance layout by casting to
+      // dynamic_parameter_impl first. Casting directly to
+      // dynamic_local_accessor_impl would result in an incorrect pointer.
+      auto *DynParamImpl = static_cast<
+          ext::oneapi::experimental::detail::dynamic_parameter_impl *>(Ptr);
+
+      MDynamicParameters.emplace_back(DynParamImpl, Index + IndexShift);
```

Comment:
There are two functional changes in this PR (this line and line 241). Before these changes, the `handler::registerDynamicParameter` method was used to add a dynamic parameter. The `handler::registerDynamicParameter` method has two checks that might throw exceptions. Since the `processArg` function is moved from `handler` to the `KernelData` we cannot use `handler::registerDynamicParameter`. I also cannot move the `registerDynamicParameter` to the `KernelData` class because it requires to access Queue and Graph from the handler.